datafusion/core/src/execution/memory_pool/pool.rs

```diff
@@ -0,0 +1,280 @@
+#[derive(Debug)]
+pub struct GreedyMemoryPool {
+    pool_size: usize,
+    used: AtomicUsize,
+}
+
+impl GreedyMemoryPool {
+    /// Allocate up to `limit` bytes
+    pub fn new(pool_size: usize) -> Self {
+        Self {
+            pool_size,
+            used: AtomicUsize::new(0),
+        }
+    }
+}
+
+impl MemoryPool for GreedyMemoryPool {
+    fn grow(&self, _allocation: &TrackedAllocation, additional: usize) {
+        self.used.fetch_add(additional, Ordering::Relaxed);
+    }
+
+    fn shrink(&self, _allocation: &TrackedAllocation, shrink: usize) {
+        self.used.fetch_sub(shrink, Ordering::Relaxed);
+    }
+
+    fn try_grow(&self, allocation: &TrackedAllocation, additional: usize) -> Result<()> {
+        self.used
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |used| {
+                let new_used = used + additional;
+                (new_used <= self.pool_size).then_some(new_used)
+            })
+            .map_err(|used| {
+                insufficient_capacity_err(allocation, additional, self.pool_size - used)
+            })?;
+        Ok(())
+    }
+
+    fn allocated(&self) -> usize {
+        self.used.load(Ordering::Relaxed)
+    }
+}
+
+/// A [`MemoryPool`] that prevents spillable allocations from using more than
+/// an even fraction of the available memory sans any unspillable allocations
+/// (i.e. `(pool_size - unspillable_memory) / num_spillable_allocations`)
+///
+///    ┌───────────────────────z──────────────────────z───────────────┐
+///    │                       z                      z               │
+///    │                       z                      z               │
+///    │       Spillable       z       Unspillable    z     Free      │
+///    │        Memory         z        Memory        z    Memory     │
+///    │                       z                      z               │
+///    │                       z                      z               │
+///    └───────────────────────z──────────────────────z───────────────┘
+///
+/// Unspillable memory is allocated in a first-come, first-serve fashion
+#[derive(Debug)]
+pub struct FairSpillPool {
+    /// The total memory limit
+    pool_size: usize,
+
+    state: Mutex<FairSpillPoolState>,
+}
+
+#[derive(Debug)]
+struct FairSpillPoolState {
+    /// The number of allocations that can spill
+    num_spill: usize,
+
+    /// The total amount of memory allocated that can be spilled
+    spillable: usize,
+
+    /// The total amount of memory allocated by consumers that cannot spill
+    unspillable: usize,
+}
+
+impl FairSpillPool {
+    /// Allocate up to `limit` bytes
+    pub fn new(pool_size: usize) -> Self {
+        Self {
+            pool_size,
+            state: Mutex::new(FairSpillPoolState {
+                num_spill: 0,
+                spillable: 0,
+                unspillable: 0,
+            }),
+        }
+    }
+}
+
+impl MemoryPool for FairSpillPool {
+    fn allocate(&self, options: &AllocationOptions) {
```

Comment:
I spent some time convincing myself this would work correctly if there are multiple SortExecs in the plan such that one would completely run before the other (e.g. as happens when there are multiple window functions with different partition / order by clauses)

I believe it will be fine because each `ExternalSorter` creates the potentially spilling allocation allocation upon construction, though I am still not 100% sure.

What would you think about moving the construction of `TrackedAllocation` into `MemoryPool`, something like 

```rust
/// Create a new tracked allocation with the MemoryManager. Subsequent requests
/// can grow or shrink the memory allocated to this allocation.
///
/// The allocation is automatically deregistered on drop
fn register(self: &Arc<Self>, options: AllocationOptions) -> TrackedAllocation;
```

That way the intent would be clearer that all ExternalSort instances register themselves with the MemoryManager on creation, and part of registering would increase the `num_spill` count.

I think having the MemoryPool create TrackedAllocations makes more logical sense as they are so tightly bound and you need a MemoryPool for a tracked allocation anyways

Reply:
> I believe it will be fine because each ExternalSorter creates the potentially spilling allocation allocation upon construction, though I am still not 100% sure.

Now that ExecutionPlan is sync I don't think this is an issue, the streams are all created prior to execution starting. Yes theoretically 

> That way the intent would be clearer that all ExternalSort instances register themselves with the MemoryManager on creation, and part of registering would increase the num_spill count.

Perhaps I'm being stupid but I don't see a tangible difference between `TrackedAllocation::new(pool, options)` and `pool.register(options)`, you're just reordering the parameters? More specifically in order for `register` to be implemented, you still need a public `TrackedAllocation::new` so you're just adding an indirection?

Reply:
I see -- I agree keeping `TrackedAllocation::new` makes sense. 

I guess I was getting confused by the fact that `allocate` doesn't actually `allocate` any memory, but instead it really registers a `TrackedAllocation` with the memory manager. Likewise `free` doesn't free any memory

Could we somehow make the relationship between `MemoryManager`,  `AllocationOptions` and `TrackedAllocation` clearer

What about changing some names:

```
MemoryManager::allocate --> MemoryManager::register
MemoryManager::free --> MemoryManager::deregister
```

Reply:
I've renamed `TrackedAllocation` to `MemoryReservation` and `AllocationOptions` to `MemoryConsumer`. I also tweaked the registration API to fit with this. PTAL